Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m

```diff
@@ -0,0 +1,178 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "ConversationViewLayout.h"
+#import "UIView+OWS.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ConversationViewLayout ()
+
+@property (nonatomic) CGSize contentSize;
+@property (nonatomic, readonly) NSMutableDictionary<NSNumber *, UICollectionViewLayoutAttributes *> *itemAttributesMap;
+
+// This dirty flag may be redundant with logic in UICollectionViewLayout,
+// but it can't hurt and it ensures that we can safely & cheaply call
+// prepareLayout from view logic to ensure that we always have a¸valid
+// layout without incurring any of the (great) expense of performing an
+// unnecessary layout pass.
+@property (nonatomic) BOOL hasLayout;
```

Comment:
> I think this boolean might be a sign that we're doing something weird here.

`UICollectionViewLayout` can be much simpler and more performant if the cells being laid out are independent of each other, e.g. they all have the same size.  

In our case, every single cell's position depends on the position of the previous cells.  It's far more efficient to layout all of the cells once and cache their layout.

As noted, the `hasLayout` might be redundant with the invalidation logic in the base class but I wanted to be really explicit about using a dirty flag here since we directly call `prepareLayout` in many places in the view where we want to synchronously ensure that layout is prepared.  I don't know of a better way to do that.

> I don't know that this is wrong but it seems unconventional, so I'm a little curious - why are we setting the contentSize in prepareLayout?

Same reason - `contentSize` is expensive to compute, we should cache it.